cruise-control-core/src/main/java/com/linkedin/cruisecontrol/monitor/sampling/MetricSample.java

```diff
@@ -0,0 +1,125 @@
+   *
+   * When the update is from metric fetcher, it does not override the user specified value.
+   *
+   * @param info        The {@link MetricInfo}
+   * @param sampleValue the sample value.
+   */
+  public void record(MetricInfo info, double sampleValue) {
+    if (_sampleTime >= 0) {
+      throw new IllegalStateException("The metric sample has been closed.");
+    }
+
+    Double origValue = _metrics.putIfAbsent(info.id(), sampleValue);
+    if (origValue != null) {
+      throw new IllegalStateException("Trying to record sample value " + sampleValue + " for " + info.name() +
+                                          ", but there is already a value " + origValue + " recorded.");
+    }
+  }
+
+  /**
+   * Record a sample value for the given resource type.
+   * This is a package private function which allows metric fetcher to override the metric if necessary.
+   * Currently it is only used when user enables auto cluster model coefficient training.
+   *
+   * When the update is from metric fetcher, it does not override the user specified value.
+   *
+   * @param name        The resource type.
+   * @param sampleValue the sample value.
+   */
+  public void record(String name, double sampleValue, MetricDef metricDef) {
+    if (_sampleTime >= 0) {
+      throw new IllegalStateException("The metric sample has been closed.");
+    }
+
+    Double origValue = _metrics.putIfAbsent(metricDef.metricInfo(name).id(), sampleValue);
+    if (origValue != null) {
+      throw new IllegalStateException("Trying to record sample value " + sampleValue + " for " + name +
+                                          ", but there is already a value " + origValue + " recorded.");
+    }
+  }
+
+  /**
+   * Get the entity this metric sample is corresponding to.
+   */
+  public E entity() {
+    return _entity;
+  }
+
+  /**
+   * The time this sample was taken.
+   */
+  public long sampleTime() {
+    return _sampleTime;
+  }
+
+  /**
+   * The metric for the specified resource.
+   */
+  public Double metricFor(int metricId) {
+    return _metrics.getOrDefault(metricId, Double.NaN);
+  }
+
+  /**
+   * @return all the metrics.
+   */
+  public Map<Integer, Double> allMetrics() {
+    return Collections.unmodifiableMap(_metrics);
+  }
+
+  /**
+   * Close this metric sample. The timestamp will be used to determine which snapshot the metric sample will be in.
+   */
+  public void close(long closingTime) {
+    if (closingTime < 0) {
+      throw new IllegalArgumentException("The closing time cannot be negative.");
+    }
+
+    if (_sampleTime < 0) {
+      _sampleTime = closingTime;
+    }
+  }
+
+  /**
+   * Validate the metric sample.
+   */
+  public boolean isValid(MetricDef metricDef) {
+    return _metrics.size() == metricDef.size();
+  }
+
+  @Override
+  public String toString() {
+    return String.format("(entity=%s,metrics=%s,sampleTime=%d)", _entity, _metrics, _sampleTime);
```

Comment:
Since this depends on many toString implementations this may generate a very large string.  In some cases that is fine.  But this may become problematic when debugging in intellij which automatically toStrings everything.